tempesta_fw/http_msg.c

```diff
@@ -475,21 +475,65 @@ __hdr_sub(TfwHttpMsg *hm, char *name, size_t n_len, char *val, size_t v_len,
 			{ .ptr = name,	.len = n_len },
 			{ .ptr = ": ",	.len = 2 },
 			{ .ptr = val,	.len = v_len },
-			{ .ptr = "\r\n", .len = 2 }
+			{ .ptr = "\r\n",.len = 2 } /* may be adjusted */
 		},
 		.len = n_len + v_len + 4,
 		.flags = 4 << TFW_STR_CN_SHIFT
 	};
+	TfwStr *hlc = TFW_STR_LAST(&hdr);
 
-	if (!TFW_STR_DUP(orig_hdr) && hdr.len <= orig_hdr->len) {
-		/* Rewrite the header in-place. */
-		if (ss_skb_cutoff_data(orig_hdr, hdr.len, 2))
-			return TFW_BLOCK;
-		if (tfw_strcpy(orig_hdr, &hdr))
-			return TFW_BLOCK;
-		return 0;
+	if (TFW_STR_DUP(orig_hdr))
+		goto slow_path;
+
+	BUG_ON(!tfw_str_eolen(orig_hdr));
+
+	/*
+	 * We are trying to keep line endings as they were received, so
+	 * adjustment of @hdr's last chunk is needed in case of single LF.
+	 */
+	if (tfw_str_eolen(orig_hdr) == 1)
+		++hlc->ptr, --hlc->len, --hdr.len;
+
+	/*
+	 * In general, there are 5 possible layouts for @hdr and @orig_hdr. And
+	 * in case of single LF used as EOL the 3-rd case is degenerated.
+	 *
+	 *   orig_hdr     | Header: Value | EOL    =  < LF | CRLF >
+	 *   ------------- --------------- ---------------------------------
+	 *
+	 *        hdr     | Foo: BarEOL.. |        1) cutoff, fast_path
+	 *        hdr     | Foo: ..BarEOL |        2) cutoff, fast_path
+	 *
+	 *        hdr     | Foo: ...BarEO | L      3) slow_path (CRLF only)
+	 *        hdr     | Foo: .....Bar | EOL    4) reuse EOL, fast_path
+	 *
+	 *        hdr     | Foo: ......Ba | rEOL   5) slow_path
+	 */
+	if (hdr.len > orig_hdr->len) {
```

Comment:
I am not sure I follow. So, we have:
- `orig_hdr = { 10 } + CRLF` -- that makes 10 bytes that we can copy.
- `hdr = { 9 + CRLF }` -- That makes 9 bytes (without `CRLF`) that we need to copy.
So it appears that we can copy these 9 bytes (without CRLF), and cut off 1 byte between the end of the substitute string and the original CRLF. Am I missing something here?

It also appears that we don't need to remove the tail at all in the fast path. Also, that means that we need to deal with the substitute string without EOL in the fast path as EOL is the same for both strings. The code sample above is not correct in that regard.